server/src/main/java/org/elasticsearch/search/rank/rrf/RRFRankShardContext.java

```diff
@@ -0,0 +1,158 @@
+        super(queries, size, from);
+        this.windowSize = windowSize;
+        this.rankConstant = rankConstant;
+    }
+
+    @Override
+    public void executeQueries(SearchContext searchContext) {
+        try {
+            QuerySearchResult querySearchResult = searchContext.queryResult();
+            RRFRankSearchContext rrfRankSearchContext = new RRFRankSearchContext(searchContext);
+
+            // run the combined boolean query for suggest, total hits, or aggregations
+            // otherwise mark top docs as empty
+            if (searchContext.suggest() != null
+                || searchContext.trackTotalHitsUpTo() != TRACK_TOTAL_HITS_DISABLED
+                || searchContext.aggregations() != null) {
+                QueryPhase.executeInternal(rrfRankSearchContext);
+            } else {
+                searchContext.queryResult()
+                    .topDocs(
+                        new TopDocsAndMaxScore(
+                            new TopDocs(new TotalHits(0, TotalHits.Relation.EQUAL_TO), Lucene.EMPTY_SCORE_DOCS),
+                            Float.NaN
+                        ),
+                        new DocValueFormat[0]
+                    );
+            }
+
+            List<TopDocs> rrfRankResults = new ArrayList<>();
+            rrfRankSearchContext.windowSize(windowSize);
+            boolean searchTimedOut = querySearchResult.searchTimedOut();
+            long serviceTimeEWMA = querySearchResult.serviceTimeEWMA();
+            int nodeQueueSize = querySearchResult.nodeQueueSize();
+
+            // run each of the rrf queries
+            for (Query query : queries) {
+                // if a search timeout occurs, exit with partial results
+                if (searchTimedOut) {
+                    break;
+                }
+                rrfRankSearchContext.rrfRankQuery(query);
+                QueryPhase.executeInternal(rrfRankSearchContext);
+                QuerySearchResult rrfQuerySearchResult = rrfRankSearchContext.queryResult();
+                rrfRankResults.add(rrfQuerySearchResult.topDocs().topDocs);
+                serviceTimeEWMA += rrfQuerySearchResult.serviceTimeEWMA();
+                nodeQueueSize = Math.max(nodeQueueSize, rrfQuerySearchResult.nodeQueueSize());
+                searchTimedOut = rrfQuerySearchResult.searchTimedOut();
+            }
+
+            sort(rrfRankResults, querySearchResult);
+
+            // record values relevant to all queries
+            querySearchResult.searchTimedOut(searchTimedOut);
+            querySearchResult.serviceTimeEWMA(serviceTimeEWMA);
+            querySearchResult.nodeQueueSize(nodeQueueSize);
+        } catch (QueryPhaseExecutionException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new QueryPhaseExecutionException(searchContext.shardTarget(), "Failed to execute main query", e);
+        }
+    }
+
+    protected void sort(List<TopDocs> rrfRankResults, QuerySearchResult querySearchResult) {
```

Comment:
I think there may be a path forward here as well, but I would like to change this when working directly in a branch with a plugin to see how it may best fit.